jiant/huggingface_transformers_interface/modules.py

```diff
@@ -355,26 +354,83 @@ def forward(self, sent: Dict[str, torch.LongTensor], task_name: str = "") -> tor
         return self.prepare_output(lex_seq, hidden_states, input_mask)
 
     def get_pretrained_lm_head(self):
-        model_with_lm_head = pytorch_transformers.RobertaForMaskedLM.from_pretrained(
+        model_with_lm_head = transformers.RobertaForMaskedLM.from_pretrained(
             self.input_module, cache_dir=self.cache_dir
         )
         lm_head = model_with_lm_head.lm_head
         lm_head.predictions.decoder.weight = self.model.embeddings.word_embeddings.weight
         return nn.Sequential(lm_head, nn.LogSoftmax(dim=-1))
 
 
-class XLNetEmbedderModule(PytorchTransformersEmbedderModule):
```

Comment:
This all looks reasonable, though it's frustrating that we can't reuse more of this code.

Could you highlight the pieces of this that you had to write yourself?

Reply:
Transformers differs slightly in what are the special tokens, where those special tokens are added in the input sequence, some input output interface. While their embedding modules and  language model heads can sometimes have considerable differences.
I know Huggingface has code that can swap in&out different kind of transformers, and their out most interface is shared (sometimes by using dummy optional inputs, which is confusing). But since we need to support things like using only word embedding and mixing inner representation, we often have to go into each transformer model and dealing with their internal differences.
Overall, I think there isn't much space for further simplification without sacrificing readability.